cypress/integration/multiple-choice.test.js

```diff
@@ -0,0 +1,124 @@
+          receivedState = newState;
+        });
+      });
+      cy.getIframeBody().find("input[value='id1']").click().then(() => {
+        expect(receivedState).eql({ selectedChoiceIds: [ "id1" ] });
+      });
+
+      cy.getIframeBody().find("input[value='id2']").click().then(() => {
+        expect(receivedState).eql({ selectedChoiceIds: [ "id2" ] });
+      });
+    });
+  });
+
+  context("Authoring view", () => {
+    it("handles pre-existing authored state", () => {
+      cy.window().should("have.property", "phone");
+      cy.window().then(window => {
+        window.phone.post("initInteractive", {
+          mode: "authoring",
+          authoredState: {
+            version: 1,
+            prompt: "Test prompt",
+            multipleAnswers: true,
+            choices: [
+              {id: "id1", content: "Choice A", correct: true},
+              {id: "id2", content: "Choice B", correct: false},
+            ]
+          }
+        });
+      });
+      cy.getIframeBody().find("#app").should("include.text", "Prompt");
+      cy.getIframeBody().find("#app").should("include.text", "Choices");
+
+      cy.getIframeBody().find("#root_prompt").should("have.value", "Test prompt");
+      cy.getIframeBody().find("#root_multipleAnswers").should("be.checked");
+      cy.getIframeBody().find("#root_choices_0_content").should("have.value", "Choice A");
+      cy.getIframeBody().find("#root_choices_0_correct").should("be.checked");
+      cy.getIframeBody().find("#root_choices_1_content").should("have.value", "Choice B");
+    });
+
+    it("renders authoring form and sends back authored state", () => {
+      let receivedState = null;
+
+      cy.window().should("have.property", "phone");
+      cy.window().then(window => {
+        window.phone.post("initInteractive", {
+          mode: "authoring"
+        });
+        window.phone.addListener("authoredState", (newState) => {
+          receivedState = newState;
+        });
+      });
+
+      cy.getIframeBody().find("#root_prompt").type("Test prompt");
+      cy.getIframeBody().find("#root_multipleAnswers").click();
+      cy.getIframeBody().find(".btn-add").click();
+      cy.getIframeBody().find("#root_choices_0_content").clear();
+      cy.getIframeBody().find("#root_choices_0_content").type("Choice A", { force: true });
+      cy.getIframeBody().find("#root_choices_0_correct").click();
+      cy.getIframeBody().find(".btn-add").click();
+      cy.getIframeBody().find("#root_choices_1_content").clear();
+      cy.getIframeBody().find("#root_choices_1_content").type("Choice B", { force: true }).then(() => {
+        expect(receivedState.version).eql(1);
+        expect(receivedState.prompt).eql("Test prompt");
+        expect(receivedState.multipleAnswers).eql(true);
+        expect(receivedState.choices[0].content).eql("Choice A");
+        expect(receivedState.choices[0].correct).eql(true);
+        expect(receivedState.choices[1].content).eql("Choice B");
```

Comment:
My cypress is rusty. If receivedState hasn't been updated yet, and this expectation fails, will cypress automatically retry the whole block?  

I know it does that in some cases

Reply:
I think it re-runs cypress matchers like `.find` or `.get`. Not sure about these lines, but I doubt so, as it's just mocha/chai. So yes, I'm cheating here. I'm explicitly relying on the fact this promise is resolved later than iframe-phone message handler. I was testing that multiple times and it was consistent. But now I think I have an idea of how to do it better and more reliable. I'll test and push in the next commit.